modules/apps/callbacks/forwarding_test.go

```diff
@@ -0,0 +1,182 @@
+package ibccallbacks_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/suite"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/ibc-go/modules/apps/callbacks/testing/simapp"
+	callbacktypes "github.com/cosmos/ibc-go/modules/apps/callbacks/types"
+	"github.com/cosmos/ibc-go/v8/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/v8/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v8/modules/core/04-channel/types"
+	ibctesting "github.com/cosmos/ibc-go/v8/testing"
+)
+
+func init() {
+	ibctesting.DefaultTestingAppInit = SetupTestingApp
+}
+
+type CallbacksForwardingTestSuite struct {
+	suite.Suite
+
+	coordinator *ibctesting.Coordinator
+
+	chainA *ibctesting.TestChain
+	chainB *ibctesting.TestChain
+	chainC *ibctesting.TestChain
+
+	pathAtoB *ibctesting.Path
+	pathBtoC *ibctesting.Path
+}
+
+// setupChains sets up a coordinator with 3 test chains.
+func (s *CallbacksForwardingTestSuite) setupChains() {
+	s.coordinator = ibctesting.NewCoordinator(s.T(), 3)
+	s.chainA = s.coordinator.GetChain(ibctesting.GetChainID(1))
+	s.chainB = s.coordinator.GetChain(ibctesting.GetChainID(2))
+	s.chainC = s.coordinator.GetChain(ibctesting.GetChainID(3))
+
+	s.pathAtoB = ibctesting.NewTransferPath(s.chainA, s.chainB)
+	s.pathBtoC = ibctesting.NewTransferPath(s.chainB, s.chainC)
+}
+
+func (s *CallbacksForwardingTestSuite) SetupTest() {
+	s.setupChains()
+
+	s.pathAtoB.Setup()
+	s.pathBtoC.Setup()
+}
+
+func TestIBCCallbacksForwardingTestsuite(t *testing.T) {
+	suite.Run(t, new(CallbacksForwardingTestSuite))
+}
+
+// TestForwardingWithMemoCallback tests that, when forwarding a packet with memo from A to B to C,
+// the callback is executed only on the final hop.
+// NOTE: this does not test the full forwarding behaviour (assert on amounts, packets, acks etc)
+// as this is covered in other forwarding tests.
+func (suite *CallbacksForwardingTestSuite) TestForwardingWithMemoCallback() {
+	testCases := []struct {
+		name                   string
+		testMemo               string
+		expCallbackMapOnChainB map[callbacktypes.CallbackType]int
+		expCallbackMapOnChainC map[callbacktypes.CallbackType]int
+	}{
+		{
+			name:                   "no memo",
+			expCallbackMapOnChainB: map[callbacktypes.CallbackType]int{},
+			expCallbackMapOnChainC: map[callbacktypes.CallbackType]int{},
+		},
+		{
+			name:                   "recv callback",
```

Comment:
It's still not super clear to me what's the mapping between Memo content and which callbacks are executed. Does anyone have a pointer to learn this better?

Reply:
the main thing is that the special keys in the callback data specify on which chain the callback should be executed.

Which is why that in this test case, it is only being executed on the `dest` chain. and not the `src` (which would be chain B in this case)